### PR TITLE
[OptApp] Restructure 11 - Add helmholtz solver

### PR DESCRIPTION
**Pull request by sunethwarna**
**📝 Description**
This PR adds a seperate helmholz solver which is a standalone FEM solver. It is then used in the `HelmholtzTransformationTechnique` as a filtering method for sensitivities in `OptimizationApplication`

Todo:
- [ ] Extend to use `Trilinos` for MPI support
- [ ] Generalize the solver to support `double` and `array3` variables seamlessly.
- [ ] Add a test

**🆕 Changelog**
- Add `HelmholtzSolver`
- Add `HelmholtzAnalysis`
- Add `HelmholtzTransformationTechnique`

### CODE REVIEW

File: applications/OptimizationApplication/python_scripts/helmholtz_analysis.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    def _CreateSolver(self):
+        return HelmholtzSolver(self.model, self.project_parameters["solver_settings"])
+
+    def _GetOrderOfProcessesInitialization(self):
+        # The list of processes will contain a list with each individual process already constructed (boundary conditions, initial conditions and gravity)
+        # Note 1: gravity is constructed first. Outlet process might need its information.
+        # Note 2: initial conditions are constructed before BCs. Otherwise, they may overwrite the BCs information.
+        return ["initial_conditions_process_list",
+                "boundary_conditions_process_list",
+                "auxiliar_process_list"]
```
Comment (FMeister) [inline applications/OptimizationApplication/python_scripts/helmholtz_analysis.py:19]:
```suggestion
                "auxiliary_process_list"]
```
Typo

File: applications/OptimizationApplication/python_scripts/solvers/helmholtz_solver.py [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+        self.main_model_part = self.model.CreateModelPart(model_part_name)
+
+        self.main_model_part.ProcessInfo[Kratos.DOMAIN_SIZE] = self.reference_model_part[Kratos.DOMAIN_SIZE]
+
+    def AddVariables(self):
+        # we add variables to the reference model part
+        # then these will be copied to the helmholtz model part
+        # with the connectivity preserve modeller
+        self.reference_model_part.AddNodalSolutionStepVariable(KratosOA.HELMHOLTZ_SOURCE_DENSITY)
+        self.reference_model_part.AddNodalSolutionStepVariable(KratosOA.HELMHOLTZ_VAR_DENSITY)
```
Comment (FMeister) [inline applications/OptimizationApplication/python_scripts/solvers/helmholtz_solver.py:37]: Name of the Kratos variable:"VAR" does not sound very descriptive. Should it mean "variance"?

File: applications/OptimizationApplication/python_scripts/solvers/helmholtz_solver.py [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+
+    def FixDofs(self, input_container_variable_data_holder: ContainerVariableDataHolderUnion):
+        self.__CheckInputContainer(input_container_variable_data_holder)
+
+        # fix the blocked flag
+        input_model_part: Kratos.ModelPart = input_container_variable_data_holder.GetModelPart()
+        Kratos.VariableUtils().SetFlag(Kratos.BLOCKED, True, input_model_part.Nodes)
+        Kratos.VariableUtils().ApplyFixity(KratosOA.HELMHOLTZ_VAR_DENSITY, True, input_model_part.Nodes)
+
+        # get the historical contaier values and set it
```
Comment (FMeister) [inline applications/OptimizationApplication/python_scripts/solvers/helmholtz_solver.py:221]:
```suggestion
        # get the historical container values and set it
```
Typo

File: applications/OptimizationApplication/python_scripts/solvers/helmholtz_solver.py [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+        historical_container_variable_data_holder = KratosOA.HistoricalContainerVariableDataHolder(input_model_part)
+        self.ConvertSolverContainer(input_container_variable_data_holder, historical_container_variable_data_holder)
+        historical_container_variable_data_holder.AssignDataToContainerVariable(KratosOA.HELMHOLTZ_VAR_DENSITY)
+
+    def GetData(self, output_container_variable_data_holder: ContainerVariableDataHolderUnion):
+        self.__CheckInputContainer(output_container_variable_data_holder)
+
+        histoical_container_variable_data_holder = KratosOA.HistoricalContainerVariableDataHolder(output_container_variable_data_holder.GetModelPart())
```
Comment (FMeister) [inline applications/OptimizationApplication/python_scripts/solvers/helmholtz_solver.py:229]:
```suggestion
        historical_container_variable_data_holder = KratosOA.HistoricalContainerVariableDataHolder(output_container_variable_data_holder.GetModelPart())
```
Typo

File: applications/OptimizationApplication/python_scripts/solvers/helmholtz_solver.py [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+        histoical_container_variable_data_holder.ReadDataFromContainerVariable(KratosOA.HELMHOLTZ_VAR_DENSITY)
```
Comment (FMeister) [inline applications/OptimizationApplication/python_scripts/solvers/helmholtz_solver.py:230]:
```suggestion
        historical_container_variable_data_holder.ReadDataFromContainerVariable(KratosOA.HELMHOLTZ_VAR_DENSITY)
```
Typo

File: applications/OptimizationApplication/python_scripts/transformation_techniques/helmholtz_transformation_technique.py [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+
+        parameters.ValidateAndAssignDefaults(default_parameters)
+        self.model = model
+        self.filter_radius = parameters["filter_radius"].GetDouble()
+        self.helmholtz_execution_policy_decorator: ExecutionPolicyDecorator = optimization_info.GetOptimizationProcess(ExecutionPolicyDecorator, parameters["helmholtz_execution_policy_name"].GetString())
+        self.fixed_model_part_names = parameters["fixed_model_part_names"].GetStringArray()
+
+        execution_policy: SteppingAnalysisExecutionPolicy = self.helmholtz_execution_policy_decorator.GetExecutionPolicy()
+        if not isinstance(execution_policy, SteppingAnalysisExecutionPolicy):
+            raise RuntimeError(f"Helmholtz tranformation requires the Helmholtz analysis to be in a SteppingAnalysisExecutionPolicy. The provided execution policy \"{self.helmholtz_execution_policy_decorator.GetExecutionPolicyName()}\" is of type {execution_policy.__class__.__name__}.")
```
Comment (FMeister) [inline applications/OptimizationApplication/python_scripts/transformation_techniques/helmholtz_transformation_technique.py:29]:
```suggestion
            raise RuntimeError(f"Helmholtz transformation requires the Helmholtz analysis to be in a SteppingAnalysisExecutionPolicy. The provided execution policy \"{self.helmholtz_execution_policy_decorator.GetExecutionPolicyName()}\" is of type {execution_policy.__class__.__name__}.")
```
Typo

File: applications/OptimizationApplication/python_scripts/transformation_techniques/helmholtz_transformation_technique.py [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+
+        self.helmholtz_analysis: HelmholtzAnalysis = execution_policy.analysis
+        if not isinstance(self.helmholtz_analysis, HelmholtzAnalysis):
+            raise RuntimeError(f"Helmholtz tranformation requires the SteppingAnalysisExecutionPolicy to have a Helmholtz analysis. The provided analysis is of type \"{self.helmholtz_execution_policy_decorator.GetExecutionPolicyName()}\" is of type {self.helmholtz_analysis.__class__.__name__}.")
```
Comment (FMeister) [inline applications/OptimizationApplication/python_scripts/transformation_techniques/helmholtz_transformation_technique.py:33]:
```suggestion
            raise RuntimeError(f"Helmholtz transformation requires the SteppingAnalysisExecutionPolicy to have a Helmholtz analysis. The provided analysis is of type \"{self.helmholtz_execution_policy_decorator.GetExecutionPolicyName()}\" is of type {self.helmholtz_analysis.__class__.__name__}.")
```
Typo

File: applications/OptimizationApplication/python_scripts/transformation_techniques/helmholtz_transformation_technique.py [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+        radius_container.SetDataForContainerVariableToZero(KratosOA.HELMHOLTZ_RADIUS_DENSITY)
+        radius_container += self.filter_radius
+        radius_container.AssignDataToContainerVariable(KratosOA.HELMHOLTZ_RADIUS_DENSITY)
+
+        container_model_part: Kratos.ModelPart = container_variable_data_holder.GetModelPart()
+
+        nodal_values = KratosOA.HistoricalContainerVariableDataHolder(container_model_part)
+        self.helmholtz_solver.ConvertSolverContainer(container_variable_data_holder, nodal_values)
+
+        mass_conribution = KratosOA.HistoricalContainerVariableDataHolder(container_model_part)
```
Comment (FMeister) [inline applications/OptimizationApplication/python_scripts/transformation_techniques/helmholtz_transformation_technique.py:60]:
```suggestion
        mass_contribution = KratosOA.HistoricalContainerVariableDataHolder(container_model_part)
```
Typo

File: applications/OptimizationApplication/python_scripts/transformation_techniques/helmholtz_transformation_technique.py [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+        KratosOA.ContainerVariableDataHolderUtils.ComputeVariableDataHolderProductWithEntityMatrix(mass_conribution, nodal_values, KratosOA.HELMHOLTZ_MASS_MATRIX, self.helmholtz_solver.GetComputingModelPart().Elements)
```
Comment (FMeister) [inline applications/OptimizationApplication/python_scripts/transformation_techniques/helmholtz_transformation_technique.py:61]:
```suggestion
        KratosOA.ContainerVariableDataHolderUtils.ComputeVariableDataHolderProductWithEntityMatrix(mass_contribution, nodal_values, KratosOA.HELMHOLTZ_MASS_MATRIX, self.helmholtz_solver.GetComputingModelPart().Elements)
```
Typo

File: applications/OptimizationApplication/python_scripts/transformation_techniques/helmholtz_transformation_technique.py [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+
+        self.helmholtz_solver.SetData(mass_conribution)
```
Comment (FMeister) [inline applications/OptimizationApplication/python_scripts/transformation_techniques/helmholtz_transformation_technique.py:63]:
```suggestion
        self.helmholtz_solver.SetData(mass_contribution)
```
Typo

**Review by FMeister — COMMENTED** [review]
I think there should be a separate folder for analysis scripts.

File: applications/OptimizationApplication/python_scripts/helmholtz_analysis.py (hunk 1 above)
Comment (sunethwarna) [reply applications/OptimizationApplication/python_scripts/helmholtz_analysis.py:19]: This is a known typo in Kratos, and unless it is changed from the core, we cannot change it.

File: applications/OptimizationApplication/python_scripts/solvers/helmholtz_solver.py (hunk 2 above)
Comment (sunethwarna) [reply applications/OptimizationApplication/python_scripts/solvers/helmholtz_solver.py:37]: These variable names will be changed to generic names accordingly.

**Comment by sunethwarna** [comment]
The `element_type` of the solver needs to be specified by the control.

File: applications/OptimizationApplication/python_scripts/transformation_techniques/helmholtz_transformation_technique.py [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+import KratosMultiphysics.OptimizationApplication as KratosOA
+from KratosMultiphysics.OptimizationApplication.transformation_techniques.transformation_technique import TransformationTechnique
+from KratosMultiphysics.OptimizationApplication.optimization_info import OptimizationInfo
+from KratosMultiphysics.OptimizationApplication.utilities.helper_utilities import ContainerVariableDataHolderUnion
+from KratosMultiphysics.OptimizationApplication.execution_policies.stepping_analysis_execution_policy import SteppingAnalysisExecutionPolicy
+from KratosMultiphysics.OptimizationApplication.utilities.execution_policy_decorator import ExecutionPolicyDecorator
+from KratosMultiphysics.OptimizationApplication.helmholtz_analysis import HelmholtzAnalysis
+from KratosMultiphysics.OptimizationApplication.solvers.helmholtz_solver import HelmholtzSolver
+
+class HelmholtzTransformationTechnique(TransformationTechnique):
```
Comment (sunethwarna) [inline applications/OptimizationApplication/python_scripts/transformation_techniques/helmholtz_transformation_technique.py:11]: this goes under `filtering`

**Comment by sunethwarna** [comment]
Closing since a better implementation was introduced and now in master.

**Pull request closed without merging**